python/ray/util/client/common.py

```diff
@@ -387,7 +396,146 @@ class ClientServerHandle:
     logs_servicer: ray_client_pb2_grpc.RayletLogStreamerServicer
     grpc_server: grpc.Server
 
+    def stop(self, grace):
+        self.grpc_server.stop(grace)
+        self.data_servicer.stopped.set()
+
     # Add a hook for all the cases that previously
     # expected simply a gRPC server
     def __getattr__(self, attr):
         return getattr(self.grpc_server, attr)
+
+
+def _get_client_id_from_context(context: Any, logger: logging.Logger) -> str:
+    """
+    Get `client_id` from gRPC metadata. If the `client_id` is not present,
+    this function logs an error and sets the status_code.
+    """
+    metadata = {k: v for k, v in context.invocation_metadata()}
+    client_id = metadata.get("client_id") or ""
+    if client_id == "":
+        logger.error("Client connecting with no client_id")
+        context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+    return client_id
+
+
+def _should_replace_cache(curr_id: int, old_id: int) -> bool:
+    """
+    We should only replace cache entries with the responses for newer IDs.
+    Most of the time newer IDs will be the ones with higher value, except when
+    the req_id counter rolls over. We check for this case by checking the
+    distance between the two IDs. If the distance is significant, then it's
+    likely that the req_id counter rolled over, and the smaller id should
+    still be used to replace the one in cache.
+    """
+    if curr_id > old_id:
+        return True
+    diff = old_id - curr_id
+    return diff > (INT32_MAX // 2)
+
+
+class ReplayCache:
+    """
+    Cache for blocking method calls. Needed to prevent retried requests from
+    being applied multiple times on the server, for example when the client
+    disconnects.
+
+    Note that no clean up logic is used, the last response for each thread
+    will always be remembered, so at most the cache will hold N entries,
+    where N is the number of threads on the client side. This relies on the
+    assumption that a thread will not make a new blocking request until it has
+    received a response for a previous one, at which point it's safe to
+    overwrite the old response.
```

Comment:
For async requests, we use streaming RPC to have serialized order. It might be easier to cache a suffix of requests in the data channel instead of per thread. We can chat offline about it later.

Reply:
Suggested logic now handled in OrderedResponseCache below

Reply:
great discussion.